### PR TITLE
Fixes dictionary size change in direct runners

### PR DESCRIPTION
**Pull request by lrakla**
Addresses #27501 by adding a `self._lock` in `shutdown_inactive_bundle_processors` similar to other functions in sdk_worker.py as suggested by @tvalentyn 
Before
``` python
def shutdown_inactive_bundle_processors():
      # type: () -> None
      for descriptor_id, last_access_time in self.last_access_times.items():
        if (time.time() - last_access_time >
            DEFAULT_BUNDLE_PROCESSOR_CACHE_SHUTDOWN_THRESHOLD_S):
          BundleProcessorCache._shutdown_cached_bundle_processors(
              self.cached_bundle_processors[descriptor_id]) 
```

After 
``` python
def shutdown_inactive_bundle_processors():
      # type: () -> None
      with self._lock:
          for descriptor_id, last_access_time in self.last_access_times.items():
            if (time.time() - last_access_time >
                DEFAULT_BUNDLE_PROCESSOR_CACHE_SHUTDOWN_THRESHOLD_S):
              BundleProcessorCache._shutdown_cached_bundle_processors(
                  self.cached_bundle_processors[descriptor_id]) 
```

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by lrakla** [comment]
R: @tvalentyn

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 1]
```diff
@@ -583,11 +583,12 @@ def _schedule_periodic_shutdown(self):
     # type: () -> None
     def shutdown_inactive_bundle_processors():
       # type: () -> None
-      for descriptor_id, last_access_time in self.last_access_times.items():
-        if (time.time() - last_access_time >
-            DEFAULT_BUNDLE_PROCESSOR_CACHE_SHUTDOWN_THRESHOLD_S):
-          BundleProcessorCache._shutdown_cached_bundle_processors(
-              self.cached_bundle_processors[descriptor_id])
+      with self._lock:
```
Comment (AnandInguva) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: from the issue, I see this is hard to reproduce but asking anyway, did you reproduce the error by any chance? If yes, can we add a test?

Comment (lrakla) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: @AnandInguva I was not sure how to reproduce it. I can take a look based on @tvalentyn comments and try reproducing it locally. This is my first contribution to the project so any guidance is appreciated. 

@tvalentyn Do you mean moving the lock definition before `self._schedule_periodic_shutdown()` ? i.e. swap lines 441 and 442?

Comment (AnandInguva) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: @lrakla ah sorry. I wasn't aware this is your first contribution. I am not sure on how to reproduce the error yet but I can try it on my local and let you know of any results and the right place to add test(if we could reproduce it).

Comment (tvalentyn) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: Thanks, @AnandInguva . @lrakla yes, that what I meant.

Comment (lrakla) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: @AnandInguva I will try it out too, and make the change @tvalentyn suggested.

Comment (AnandInguva) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: I tried few simple examples with 1000, 10000, 100000 elements by setting `direct_num_worker=4` and `direct_running_mode=multi_threading`. I haven't been able to reproduced the error.

Comment (AnandInguva) [reply sdks/python/apache_beam/runners/worker/sdk_worker.py:586]: tried it with 1. I will try with 0.1

**Comment by damccorm** [comment]
@AnandInguva @tvalentyn @lrakla what are next steps here?

**Review by tvalentyn — APPROVED** [review]

**Review by tvalentyn — APPROVED** [review]

**Review by tvalentyn — APPROVED** [review]

**Review by tvalentyn — APPROVED** [review]

**Pull request merged**